src/mlpack/core/data/split_data.hpp

```diff
@@ -17,6 +17,138 @@
+ * arma::mat testData;
+ * arma::Row<size_t> trainLabel;
+ * arma::Row<size_t> testLabel;
+ * math::RandomSeed(100); // Set the seed if you like.
+ *
+ * // Stratify the dataset into a training and test set, with 30% of the data
+ * // being held out for the test set.
+ * StratifiedSplit(input, label, trainData,
+ *                 testData, trainLabel, testLabel, 0.3);
+ * @endcode
+ *
+ * @param input Input dataset to stratify.
+ * @param inputLabel Input labels to stratify.
+ * @param trainData Matrix to store training data into.
+ * @param testData Matrix to store test data into.
+ * @param trainLabel Vector to store training labels into.
+ * @param testLabel Vector to store test labels into.
+ * @param testRatio Percentage of dataset to use for test set (between 0 and 1).
+ * @param shuffleData If true, the sample order is shuffled; otherwise, each
+ *     sample is visited in linear order. (Default true.)
+ */
+template<typename T, typename U>
+void StratifiedSplit(const arma::Mat<T>& input,
+                     const arma::Row<U>& inputLabel,
+                     arma::Mat<T>& trainData,
+                     arma::Mat<T>& testData,
+                     arma::Row<U>& trainLabel,
+                     arma::Row<U>& testLabel,
+                     const double testRatio,
+                     const bool shuffleData = true)
+{
+  /**
+   * Basic idea:
+   * Let us say we have to stratify a dataset based on labels:
+   * 0 0 0 0 0 (5 0s)
+   * 1 1 1 1 1 1 1 1 1 1 1 (11 1s)
+   *
+   * Let our test ratio be 0.2.
+   * Then, the number of 0 labels in our test set = floor(5 * 0.2) = 1.
+   * The number of 1 labels in our test set = floor(11 * 0.2) = 2.
+   *
+   * In our first pass over the dataset,
+   * We visit each label and keep count of each label in our 'labelCounts' uvec.
+   *
+   * We then take a second pass over the dataset.
+   * We now maintain an additional uvec 'testLabelCounts' to hold the label
+   * counts of our test set.
+   *
+   * In this pass, when we encounter a label we check the 'testLabelCounts' uvec
+   * for the count of this label in the test set.
+   * If this count is less than the required number of labels in the test set,
+   * we add the data to the test set and increment the label count in the uvec.
+   * If this count is equal to or more than the required count in the test set,
+   * we add this data to the train set.
+   *
+   * Based on the above steps, we get the following labels in the split set:
+   * Train set (4 0s, 9 1s)
+   * 0 0 0 0
+   * 1 1 1 1 1 1 1 1 1
+   *
+   * Test set (1 0s, 2 1s)
+   * 0
+   * 1 1
+   */
+  size_t trainIdx = 0;
+  size_t testIdx = 0;
+  size_t trainSize = 0;
+  size_t testSize = 0;
+  arma::uvec labelCounts;
+  arma::uvec testLabelCounts;
+  U maxLabel = inputLabel.max();
+
+  labelCounts.zeros(maxLabel+1);
+  testLabelCounts.zeros(maxLabel+1);
+
+  arma::uvec order =
+      arma::linspace<arma::uvec>(0, input.n_cols - 1, input.n_cols);
+
+  for (arma::uword i : order)
```

Comment:
Sorry, just one more thing... why not iterate over `inputLabel` directly?  The compiler doesn't have any way to know that `order` is ordered, and so it won't be able to make any optimizations (at the instruction level) resulting from the linear pass over `inputLabel`.  (Of course, since the actual lookup behavior is passing through the memory in a contiguous fashion, the processor itself will manage to still load things in more quickly than a random lookup.  Anyway, still better to allow the compiler the information to know it's a linear pass.)

Reply:
Right, that makes sense. I have made the change.